### ISSUE TITLE
kube-flannel: iptables segfault after upgrade to v0.13.0

### ISSUE THREAD

**Issue by x3nb63** [flannel-io/flannel#1408]
The kube-flannel DeamonSet pod fails to check rule existence on some (?) rules, which exists with -1. See log below.

When inspecting the node I find a segfault for each line with exit -1, see the other log.

## Expected Behavior

No segfault.

I have no idea what its checking here. There may be non-existing pods in my etcd database. Whatever it is, it should not segfault on it.

## Current Behavior

it segfaults every 5 to 10 minutes, on all nodes in the cluster, including master nodes.

(probably cause they all run the same docker image)

logs:

```
$ kubectl -n kube-system logs pod/kube-flannel-ds-vgsh5
...
E0201 17:13:29.425597       1 iptables.go:115] Failed to ensure iptables rules: Error checking rule existence: failed to check rule existence: running [/sbin/iptables -t nat -C POSTROUTING ! -s 10.244.0.0/16 -d 10.244.0.0/16 -j MASQUERADE --random-fully --wait]: exit status -1: 
E0201 17:17:25.111358       1 iptables.go:115] Failed to ensure iptables rules: Error checking rule existence: failed to check rule existence: running [/sbin/iptables -t filter -C FORWARD -d 10.244.0.0/16 -j ACCEPT --wait]: exit status -1: 

```
the node:

```
[root@master-node1 ~]# coredumpctl info 2200900
           PID: 2200900 (iptables)
           UID: 0 (root)
           GID: 0 (root)
        Signal: 11 (SEGV)
     Timestamp: Mon 2021-02-01 18:13:28 CET (2min 35s ago)
  Command Line: /sbin/iptables -t nat -C POSTROUTING -s 10.244.0.0/16 -d 10.244.0.0/16 -j MASQUERADE --random-fully --wait
    Executable: /sbin/xtables-nft-multi
 Control Group: /kubepods/burstable/podaecc8736-4000-4010-8a94-b49a73b56882/a726f110b905636614f762cc94114afa6810e9a2d5b1c65cdbbd4858943f4118
         Slice: -.slice
       Boot ID: a27b5949a5eb412ac95fb5c89a0871c3
    Machine ID: af7e08e4fa8e42e382da98334f96a1c5
      Hostname: master-host1
       Storage: /var/lib/systemd/coredump/core.iptables.0.a27b5949a5eb412aa95fb5c89a0871c3.2200900.1612199608000000.lz4
       Message: Process 2200900 (iptables) of user 0 dumped core.
                
                Stack trace of thread 1535:
                #0  0x00007f1392855e47 n/a (/usr/lib/libnftnl.so.11.3.0)
```

## Possible Solution

Dont know.

## Steps to Reproduce (for bugs)

Don't know as of yet. Don't see what it wants to check.

## Context

Cluster was upgraded to Kubernetes 1.20.2 and Flannel along with it.

Accidentally I upgraded to v0.13.1-rc1 first, seen these segfaults, then downgraded to v0.13.0. Did not solve it.

## Your Environment

* Flannel version: `v0.13.0`
* Backend used (e.g. vxlan or udp):  `vxlan`
* Etcd version: `3.4.13-0`
* Kubernetes version (if used): `v1.20.2`
* Operating System and version: `CentOS Linux release 8.3.2011`
* Kernel: `4.18.0-240.10.1.el8_3.x86_64`
* Docker version: `docker-ce-19.03.7-3.el7.x86_64`

**Comment by oncipriani** [comment]
Same issue here:

- Flannel version: `v0.13.0-rancher1`
- Backend used (e.g. vxlan or udp): `host-gw`
- Etcd version: `v3.4.13-rancher1`
- Kubernetes version (if used): `v1.19.6-rancher1-1`
- Operating System and version: `CentOS Linux release 8.3.2011`
- Kernel: `Linux 4.18.0-240.1.1.el8_3.x86_64` and `4.18.0-240.10.1.el8_3.x86_64`
- Docker version: `docker-ce-20.10.0-3.el8.x86_64`

**Comment by Oats87** [comment]
Interesting.

With `v0.13.0`+ of Flannel, there is auto-detection of the underlying "packet filter" instead of just using `iptables` in `legacy` mode.

As given the binary that is being executed i.e. `xtables-nft-multi`, the logic has chosen `nft` mode which makes sense on `CentOS 8` but at the end of the day I see an overlap of both OS + kernel versions so I will test this to validate.

**Comment by Oats87** [comment]
I'm testing on a `CentOS Linux release 8.3.2011` system with `4.18.0-240.10.1.el8_3.x86_64` as a kernel and `libnftnl-1.1.5-4.el8.x86_64` on the underlying host and am unable to reproduce this segfault. Of course, the `libnftnl` that is used by Flannel is the one that comes from the userspace of the container, and is Alpine-provided -- in our Flannel `v0.13.0` container this should be like: `libnftnl-libs-1.1.6-r0 x86_64 {libnftnl} (GPL-2.0-or-later) [installed]`

On a system with a crashing flannel pod, would you be able to provide the output of `lsmod`? Also, the output of `rpm -q libnftnl` would be helpful as well.

**Comment by Oats87** [comment]
I'm not sure what timezone you guys are based out of; if you're close to PST I would love to try and live-debug this/collect more specific logs. I can be found on the Kubernetes Slack under the name "Chris Kim"

**Comment by x3nb63** [comment]
sorry for the delay!

its still happening. Here is an lsmod from the master node I made this issue with:

```
[root@master-node1 ~]# lsmod 
Module                  Size  Used by
ipt_REJECT             16384  0
nf_reject_ipv4         16384  1 ipt_REJECT
veth                   28672  0
vxlan                  65536  0
ip6_udp_tunnel         16384  1 vxlan
udp_tunnel             16384  1 vxlan
xt_statistic           16384  6
xt_physdev             16384  42
xt_nat                 16384  97
ip_vs_sh               16384  0
ip_vs_wrr              16384  0
ip_vs_rr               16384  0
ip_vs                 172032  6 ip_vs_rr,ip_vs_sh,ip_vs_wrr
xt_comment             16384  666
xt_mark                16384  6
xt_conntrack           16384  9
ipt_MASQUERADE         16384  4
nf_conntrack_netlink    49152  0
nft_counter            16384  694
xt_addrtype            16384  87
nft_compat             20480  1374
nft_chain_nat          16384  4
nf_nat                 45056  3 ipt_MASQUERADE,xt_nat,nft_chain_nat
nf_conntrack          172032  6 xt_conntrack,nf_nat,ipt_MASQUERADE,xt_nat,nf_conntrack_netlink,ip_vs
nf_defrag_ipv6         20480  2 nf_conntrack,ip_vs
nf_defrag_ipv4         16384  1 nf_conntrack
nf_tables             167936  3055 nft_compat,nft_counter,nft_chain_nat
nfnetlink              16384  4 nft_compat,nf_conntrack_netlink,nf_tables
br_netfilter           24576  0
bridge                192512  1 br_netfilter
stp                    16384  1 bridge
llc                    16384  2 bridge,stp
rtl8723be             106496  0
btcoexist             163840  1 rtl8723be
rtl8723_common         28672  1 rtl8723be
rtl_pci                32768  1 rtl8723be
rtlwifi                98304  4 rtl_pci,rtl8723be,btcoexist,rtl8723_common
mac80211              901120  3 rtl_pci,rtl8723be,rtlwifi
edac_mce_amd           32768  0
ccp                    98304  0
wmi_bmof               16384  0
btusb                  53248  0
btrtl                  20480  1 btusb
kvm                   798720  0
btbcm                  20480  1 btusb
btintel                24576  1 btusb
cfg80211              806912  2 rtlwifi,mac80211
uvcvideo              114688  0
bluetooth             675840  5 btrtl,btintel,btbcm,btusb
snd_hda_codec_conexant    20480  1
snd_hda_codec_generic    86016  1 snd_hda_codec_conexant
videobuf2_vmalloc      16384  1 uvcvideo
irqbypass              16384  1 kvm
videobuf2_memops       16384  1 videobuf2_vmalloc
snd_hda_codec_hdmi     69632  1
snd_hda_intel          49152  0
crct10dif_pclmul       16384  1
crc32_pclmul           16384  0
videobuf2_v4l2         28672  1 uvcvideo
ghash_clmulni_intel    16384  0
snd_intel_dspcfg       24576  1 snd_hda_intel
videobuf2_common       53248  2 videobuf2_v4l2,uvcvideo
videodev              204800  2 videobuf2_v4l2,uvcvideo
snd_hda_codec         151552  4 snd_hda_codec_generic,snd_hda_codec_conexant,snd_hda_codec_hdmi,snd_hda_intel
joydev                 24576  0
pcspkr                 16384  0
snd_hda_core           98304  5 snd_hda_codec_generic,snd_hda_codec_conexant,snd_hda_codec_hdmi,snd_hda_intel,snd_hda_codec
k10temp                16384  0
fam15h_power           16384  0
thinkpad_acpi         110592  0
sp5100_tco             16384  0
ecdh_generic           24576  1 bluetooth
snd_hwdep              16384  1 snd_hda_codec
i2c_piix4              24576  0
ftdi_sio               61440  0
libarc4                16384  1 mac80211
rtsx_pci_ms            20480  0
snd_pcm               118784  4 snd_hda_codec_hdmi,snd_hda_intel,snd_hda_codec,snd_hda_core
memstick               16384  1 rtsx_pci_ms
overlay               135168  16
snd_timer              40960  1 snd_pcm
ledtrig_audio          16384  3 snd_hda_codec_generic,snd_hda_codec_conexant,thinkpad_acpi
snd                    94208  9 snd_hda_codec_generic,snd_hda_codec_conexant,snd_hda_codec_hdmi,snd_hwdep,snd_hda_intel,snd_hda_codec,snd_timer,thinkpad_acpi,snd_pcm
rfkill                 28672  4 bluetooth,thinkpad_acpi,cfg80211
wmi                    32768  1 wmi_bmof
i2c_scmi               16384  0
soundcore              16384  1 snd
video                  49152  1 thinkpad_acpi
acpi_cpufreq           24576  0
ext4                  761856  1
mbcache                16384  1 ext4
jbd2                  131072  1 ext4
ip_tables              28672  0
xfs                  1511424  1
libcrc32c              16384  4 nf_conntrack,nf_nat,xfs,ip_vs
sr_mod                 28672  0
cdrom                  65536  1 sr_mod
sd_mod                 53248  3
sg                     40960  0
radeon               1622016  1
rtsx_pci_sdmmc         28672  0
mmc_core              172032  1 rtsx_pci_sdmmc
i2c_algo_bit           16384  1 radeon
drm_kms_helper        217088  1 radeon
ahci                   40960  2
libahci                40960  1 ahci
syscopyarea            16384  1 drm_kms_helper
sysfillrect            16384  1 drm_kms_helper
sysimgblt              16384  1 drm_kms_helper
fb_sys_fops            16384  1 drm_kms_helper
ttm                   110592  1 radeon
libata                270336  2 libahci,ahci
crc32c_intel           24576  3
drm                   557056  4 drm_kms_helper,radeon,ttm
serio_raw              16384  0
r8169                  86016  0
rtsx_pci               73728  2 rtsx_pci_sdmmc,rtsx_pci_ms
realtek                20480  1
```

and rpm

```
[root@master-node1 ~]# rpm -q libnftnl
libnftnl-1.1.5-4.el8.x86_64
```

btw: the only purpose of this machine is to provide more quorum in my cluster, doesn't run anything but kube-system (which is why I debug on it)

**Comment by Oats87** [comment]
Interesting.

Would you be able to exec into the `flannel` container and perform a `iptables` and `iptables-save` and see if either segfault?

**Comment by x3nb63** [comment]
nope, no segfault

```
bash-5.0# iptables
iptables v1.8.4 (nf_tables): no command specified
Try `iptables -h' or 'iptables --help' for more information.
bash-5.0# iptables-save | head
# Generated by iptables-save v1.8.4 on Tue Feb  9 17:33:40 2021
*nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:KUBE-MARK-DROP - [0:0]
:KUBE-MARK-MASQ - [0:0]
:KUBE-POSTROUTING - [0:0]
:KUBE-KUBELET-CANARY - [0:0]
bash-5.0#
```

**Comment by Oats87** [comment]
Interesting...

Do you have `selinux` or anything else on this system enabled?

Are you on the Kubernetes Slack? If so, can we try to troubleshoot via DM there?

**Comment by x3nb63** [comment]
selinux is off if I understand that right:

```
[root@master-node1 ~]# selinuxenabled 
[root@master-node1 ~]# echo $?
1
```

I do not have Slack nor anything similar. But am open to experiments as long as they don't kill the cluster.

**Comment by x3nb63** [comment]
I just ran the command from the last segfault manually in the `kube-flannel-ds` container shell and it did this:

```
bash-5.0# /sbin/iptables -t filter -C FORWARD -s 10.42.0.0/16 -j ACCEPT --wait
bash-5.0# echo $?
0
```

no segfault on the host. (I watch with `journalctl -f`)

**Comment by Oats87** [comment]
I massaged my system a bit and was able to generate a few coredumps.

It looks like this is happening on random calls of `iptables`, 
```
Core was generated by `/sbin/iptables -t filter -C FORWARD -d 10.42.0.0/16 -j ACCEPT --wait'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f26f12b7e47 in nftnl_rule_lookup_byindex () from /usr/lib/libnftnl.so.11
```

Doing some digging on this, it seems that CentOS is not the only distribution that is hitting this issue, namely Debian has seen it: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=951477

More investigation into this needs to be performed, but on this first-glance, it appears that this is an issue with `iptables` and `libnftnl` rather than it being a Flannel-specific issue. In the mean time, your system *should* be functional as Flannel will simply retry (and eventually succeed) at making a call.

**Comment by x3nb63** [comment]
cool, thanks.

so for now, other then filling the journal with coredumps there is no effect on the traffic in the cluster - thats good, I can wait.

is there some upstream reporting necessary? (not that I know how to describe it properly)

**Comment by rsoika** [comment]
When I install my worker nodes I use s a script which is including the following setup commands for containerd
Can this be related to the issue?

**Comment by rsoika** [comment]
Is it possible that this issue is related to #1415 ?

**Comment by x3nb63** [comment]
> When I install my worker nodes I use s a script which is including the following setup commands for containerd
> Can this be related to the issue?
>  [...]

My system runs the same modules and sysctl settings - to my knowledge they are mandated by either Kubernetes or Flannel. In any case they are necessary to have public network in containers. I'd say it is related because it has to be.

> Is it possible that this issue is related to #1415 ?

My `kube-flannel` pods run with 50Mi memory limit and I don't observe OOMkills. But note mine are still at v0.13.0.  (I don't dare testing an rc)

**Comment by rsoika** [comment]
In the mean time I switched to calico and it seems the problem is gone.
But it is also possible that it was due to the missing library 'nftables' which I have now added to all nodes.

**Comment by chunhualiu** [comment]
I am having the same issue:
```
Apr 27 14:30:06 k8s-2-62 kernel: iptables[2405300]: segfault at 88 ip 00007f05003aee47 sp 00007ffce7b8a6b8 error 4 in libnftnl.so.11.3.0[7f05003aa000+16000]
Apr 27 14:30:08 k8s-2-62 systemd-coredump[2405305]: Process 2405300 (iptables) of user 0 dumped core.#012#012Stack trace of thread 494477:#012#0  0x00007f05003aee47 n/a (/usr/lib/libnftnl.so.11.3.0)
Apr 27 14:32:11 k8s-2-62 kernel: iptables-save[2408686]: segfault at 80 ip 00007fd871860964 sp 00007ffc4a4fbbe8 error 4 in libnftnl.so.11.2.0[7fd87185b000+19000]
Apr 27 14:32:15 k8s-2-62 systemd-coredump[2408704]: Process 2408686 (iptables-save) of user 0 dumped core.#012#012Stack trace of thread 2408686:#012#0  0x00007fd871860964 n/a (/usr/lib/x86_64-linux-gnu/libnftnl.so.11.2.0)
Apr 27 14:32:43 k8s-2-62 kernel: iptables[2409989]: segfault at 80 ip 00007f6fa1ba8964 sp 00007ffe949a3088 error 4 in libnftnl.so.11.2.0[7f6fa1ba3000+19000]
Apr 27 14:32:44 k8s-2-62 systemd-coredump[2410020]: Process 2409989 (iptables) of user 0 dumped core.#012#012Stack trace of thread 2409989:#012#0  0x00007f6fa1ba8964 n/a (/usr/lib/x86_64-linux-gnu/libnftnl.so.11.2.0)
Apr 27 15:04:16 k8s-2-62 kernel: iptables-save[2442071]: segfault at 80 ip 00007ff9a67fe964 sp 00007ffe46ac9428 error 4 in libnftnl.so.11.2.0[7ff9a67f9000+19000]
Apr 27 15:04:16 k8s-2-62 systemd-coredump[2442077]: Process 2442071 (iptables-save) of user 0 dumped core.#012#012Stack trace of thread 2442071:#012#0  0x00007ff9a67fe964 n/a (/usr/lib/x86_64-linux-gnu/libnftnl.so.11.2.0)
Apr 27 15:08:21 k8s-2-62 kernel: iptables-save[2446175]: segfault at 80 ip 00007fe7fbeec964 sp 00007ffde0ec2528 error 4 in libnftnl.so.11.2.0[7fe7fbee7000+19000]
Apr 27 15:08:22 k8s-2-62 systemd-coredump[2446183]: Process 2446175 (iptables-save) of user 0 dumped core.#012#012Stack trace of thread 2446175:#012#0  0x00007fe7fbeec964 n/a (/usr/lib/x86_64-linux-gnu/libnftnl.so.11.2.0)
Apr 27 15:33:34 k8s-2-62 kernel: iptables[2470414]: segfault at 88 ip 00007f13dcdc0e47 sp 00007ffeddb003e8 error 4 in libnftnl.so.11.3.0[7f13dcdbc000+16000]
Apr 27 15:33:35 k8s-2-62 systemd-coredump[2470420]: Process 2470414 (iptables) of user 0 dumped core.#012#012Stack trace of thread 499035:#012#0  0x00007f13dcdc0e47 n/a (/usr/lib/libnftnl.so.11.3.0)
Apr 27 15:43:06 k8s-2-62 kernel: iptables[2479683]: segfault at 88 ip 00007f12c0a02e47 sp 00007fff32d72c08 error 4 in libnftnl.so.11.3.0[7f12c09fe000+16000]
Apr 27 15:43:06 k8s-2-62 systemd-coredump[2479685]: Process 2479683 (iptables) of user 0 dumped core.#012#012Stack trace of thread 499722:#012#0  0x00007f12c0a02e47 n/a (/usr/lib/libnftnl.so.11.3.0)
```

OS: CentOS 8
Kernel: 4.18.0-193.28.1.el8_2.x86_64
flanneld: v0.13.0-rancher1
rancher: v2.5.5

**Comment by rsoika** [comment]
One thing which may be an issue for Debian OS is the library 'nftables' which seems to be important in combination with IP-Tables. This lib was missing in my setup.

**Comment by ciaby** [comment]
I think I tracked down this issue.
This bug looks very much like this:
https://bugzilla.redhat.com/show_bug.cgi?id=1812261
The bug got fixed by Redhat first and then in iptables 1.8.5 upstream:
https://www.netfilter.org/projects/iptables/files/changes-iptables-1.8.5.txt
The issue apparently started with flannel 0.13.0, which is when alpine 3.12 got introduced. That means that in order to fix it we need to move to alpine 3.13 which ships with iptables 1.8.6.
I'm opening a MR :)
